src/cluster/agglomerative.rs

```diff
@@ -0,0 +1,219 @@
+//! 
+//! `MST-linkage` is an implmenetation of `MST-linkage-core` plus two post-processing steps:
+//! * Sort by distance
+//! * LABEL (aka union-find), as in Fig.5 in Müllner, 2011
+//! 
+//! 
+//! Example:
+//!
+//! ```
+//! use smartcore::linalg::naive::dense_matrix::*;
+//! use smartcore::cluster::agglomerative::*;
+//!
+//! // Iris data
+//! let x = DenseMatrix::from_2d_array(&[
+//!            &[5.1, 3.5, 1.4, 0.2],
+//!            &[4.9, 3.0, 1.4, 0.2],
+//!            &[4.7, 3.2, 1.3, 0.2],
+//!            &[4.6, 3.1, 1.5, 0.2],
+//!            &[5.0, 3.6, 1.4, 0.2],
+//!            &[5.4, 3.9, 1.7, 0.4],
+//!            &[4.6, 3.4, 1.4, 0.3],
+//!            &[5.0, 3.4, 1.5, 0.2],
+//!            &[4.4, 2.9, 1.4, 0.2],
+//!            &[4.9, 3.1, 1.5, 0.1],
+//!            &[7.0, 3.2, 4.7, 1.4],
+//!            &[6.4, 3.2, 4.5, 1.5],
+//!            &[6.9, 3.1, 4.9, 1.5],
+//!            &[5.5, 2.3, 4.0, 1.3],
+//!            &[6.5, 2.8, 4.6, 1.5],
+//!            &[5.7, 2.8, 4.5, 1.3],
+//!            &[6.3, 3.3, 4.7, 1.6],
+//!            &[4.9, 2.4, 3.3, 1.0],
+//!            &[6.6, 2.9, 4.6, 1.3],
+//!            &[5.2, 2.7, 3.9, 1.4],
+//!            ]);
+//!
+//! let primitive_clustering = PrimitiveClustering::fit(&x, 2, Default::default()).unwrap(); // Fit to data, 2 clusters
+//!
+use std::cmp;
+use crate::math::num::RealNumber;
+use smartcore::math::distance::euclidian::Euclidian;
+
+
+pub trait SAHNClustering<T: RealNumber> {
+  pub fn fit<M: Matrix<T>>(
+    data: &M,
+    k: usize,
+  ) -> Result<ClusterResult<T>, Failed>;
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct PrimitiveClustering {}
+
+impl PrimitiveClustering {
+  fn compute(&self) ->  {
+    // 1: procedure Primitive_clustering(S, d) . S: node labels, d: pairwise dissimilarities
+    // 2: N ← |S| . Number of input nodes
+    // 3: L ← [ ] . Output list
+    // 4: size[x] ← 1 for all x ∈ S
+    // 5: for i ← 0, . . . , N − 2 do
+    // 6: (a, b) ← argmin(S×S)\∆ d
+    // 7: Append (a, b, d[a, b]) to L.
+    // 8: S ← S \ {a, b}
+    // 9: Create a new node label n /∈ S.
+    // 10: Update d with the information
+    // d[n, x] = d[x, n] = Formula(d[a, x], d[b, x], d[a, b], size[a], size[b], size[x])
+    // for all x ∈ S.
+    // 11: size[n] ← size[a] + size[b]
+    // 12: S ← S ∪ {n}
+    // 13: end for
+    // 14: return L . the stepwise dendrogram, an ((N − 1) × 3)-matrix
+    // 15: end procedure
+    // (As usual, ∆ denotes the diagonal in the Cartesian product S × S.)
+  }
+}
+
+///
+/// An implementation of `PRIMITIVE_CLUSTERING`
+///
+impl SAHNClustering for PrimitiveClustering {
+
+}
+
+
+///
+/// The triple defined as (observation_1, observation_2, distance).
+/// This defines the type for the elements of the list that 
+///   represents the Stepwise Dendrogram (Müllner, 2011)
+/// 
+#[derive(Debug)]
+struct PairwiseDissimilarity<'a> {
+    node1: Vec<RealNumber>;
+    node2: Vec<RealNumber>;
+    distance: f32;
```

Comment:
Are you going to replace all occurrences of f32 by generic type later? If not this might be a problem since:
1. We'll have to convert from T to f32 every time you calculate cluster dissimilarity  
2. Precision might potentially be affected because of the rounding error

Reply:
sure, just making things work first.